### ISSUE TITLE
YOLOv8-Pose issue with hand keypoints

### ISSUE THREAD

**Issue by a-mohsen** [ultralytics/ultralytics#2092]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to train hand pose estimation (21 keypoints) on YOLOv8-Pose, but I am encountering issues during training
Everytime after few epochs, the pose accuracy begins to deteriorate until it reaches zero. Furthermore, I am continuously receiving warnings throughout the training process, specifically:
RuntimeWarning: overflow encountered in multiply
self.keypoints[..., 0] *= w
RuntimeWarning: overflow encountered in multiply
self.keypoints[..., 1] *= h

![image](https://user-images.githubusercontent.com/7348839/232645467-3b9ad47d-254a-4216-bea5-e001fe861317.png)

I have attempted to resolve the problem by implementing a number of changes, including adjusting the optimizer, modifying the LR and epoch size, changing the keypoints gain, disabling the augmentation (especially the flipping), changing the training/val splitting seed, and modifying the keypoints format to (x,y - x,y,visibility). I have even limited the keypoints annotations to 5 decimals, but the problem still persists.

I would appreciate any insight or advice regarding it.

**Comment by Laughing-q** [comment]
@a-mohsen hi, is there a public dataset of hand pose estimation? I'd like to try it and investigate this issue if possible.
We verified the the current v8 pose implementation on human-pose and facial landmark and both of them work very well. But we never tried to train a hand pose estimation model, it would be great if you can point me a public hand pose dataset that I can verify and give you feedback. :)

**Comment by a-mohsen** [comment]
Thanks @Laughing-q !
I'm using [multiviewDataset](https://github.com/ShichengChen/multiviewDataset) RGB images.

**Comment by Laughing-q** [comment]
@a-mohsen ok got it! thanks! I'll feedback once I get something.

**Comment by Laughing-q** [comment]
@a-mohsen hey, it just occurred to me that this might be related to a bug in our previous version, and it had been fixed in the latest version.
can you tell me your package version of ultralytics?

**Comment by a-mohsen** [comment]
@Laughing-q you're right!
I was using 8.0.73. After upgrading to the latest version the warnings have gone and the training goes smoothly.
Thanks a lot, I really appreciate it.

**Comment by glenn-jocher** [comment]
@a-mohsen, glad to hear that upgrading to the latest version solved your problem! We always encourage users to keep their packages up-to-date, as this will help them avoid problems and also take advantage of new features and enhancements. 
If you have any further questions, don't hesitate to reach out again.

**Comment by blackCmd** [comment]
@a-mohsen 
Hello. I want to detect the key points of a car like you do, but I'm having a lot of trouble learning.
![image](https://github.com/ultralytics/ultralytics/assets/13531174/dfad16d6-3118-4b7a-8b6d-f3d32cfd84bb)


More details can be found at the link [Here.](https://github.com/ultralytics/ultralytics/issues/3320)

What am I doing wrong? I'm struggling with learning just like you for a long time. Thank you.

**Comment by seawater668** [comment]
> @Laughing-q你说得对！ 我用的是8.0.73。升级到最新版本后，警告消失了，训练顺利进行。 非常感谢，我真的很感激。

Hello, if there are no key points in the gesture image annotation, how should the coordinates of the key points be annotated? Do you use one category for all gestures, or do you detect gesture key points in multiple categories?

**Comment by seawater668** [comment]
> @a-mohsen，很高兴听到升级到最新版本解决了您的问题！我们始终鼓励用户保持其软件包最新，因为这将帮助他们避免问题并利用新功能和增强功能。 如果您还有任何疑问，请随时再次与我们联系。

If there are no key points in the training data set annotation, how should the coordinates of the key points be annotated? Is it possible to use 2D key points for annotation without being visible or invisible?

**Comment by glenn-jocher** [comment]
@seawater668 hello, if there are no keypoints in your images, you can annotate them as negatively infinite or any other identifier that denotes absence. For your other question, the visibility of the keypoints in the images is crucial for the training. YOLOv8 supports 2D keypoints annotation including the x, y coordinates, and also the visibility flag (visible or invisible), and all three should be provided during annotation. The visibility field is an important part because it guides the model in learning when a keypoint might not be visible due to occlusions or other reasons. Hope this helps!

**Comment by seawater668** [comment]
@glenn-jocher 
It’s still a question of data annotation. If the key points I detect are 8, and there are two bboxes in a picture, one of which has 8 key points, and one of which has no key points, should I also use 8x3 all-zero key points? Is it written in a txt document?
According to the above description, the example is as follows...

1 0.31324196 0.21755578  0.10160532  0.15264371  0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000 0.31324196 0.21755578 2.000000
0 0.64766242  0.76419351  0.04947515  0.11232601 0.000000  0.000000  0.000000  0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000 0.000000  0.000000  0.000000

**Comment by glenn-jocher** [comment]
@seawater668 You've got it right. When a bounding box does not have any keypoints, you should indeed use zero vectors to signify their absence. 

Each bbox in your annotation should have a consistent set of keypoints, regardless of whether they are particularly visible or not. In situations when there are no keypoints associated with a bounding box, as in the second box of your example, writing 8x3 all-zero vectors is completely correct.

This uniform structure will be beneficial in helping your model learn that the absence of keypoints in specific bounding box scenarios is possible and not an error. 

I hope this helps, and feel free to ask if you have any additional questions.

**Comment by saqaw** [comment]
@a-mohsen sir, is that your work publicaly available . did you train hand keypoints with yolo v8. can i get trained model weight file to use/ test for my datatset

**Comment by TaylorT52** [comment]
can someone post the dataset? I haven't been able to find any hand keypoint datasets in yolov8-pose format

**Comment by glenn-jocher** [comment]
@TaylorT52 i'm sorry, but we don't provide datasets. However, you can create your own dataset by annotating images in the YOLOv8-Pose format. The [Ultralytics Docs](https://docs.ultralytics.com) provide guidance on how to prepare and format your data for training. Good luck with your project! 🚀
